### ISSUE TITLE
request user data taking >10000ms when using websockets

### ISSUE THREAD

**Issue by atkinsj** [exodus4d/pathfinder#439]
I'm not sure how to begin troubleshooting this but that seems inordinately high. Update users and update map all occur in <5ms but update user data ranges from 8,000 to 10,000ms. How do I go about troubleshooting this?

**Comment by dkrotil** [comment]
If you have single CPU system, then WebSocket server eats all your CPU cycles.

**Comment by atkinsj** [comment]
Heh, I'm running it on a quad core and it is indeed using 100% CPU. How is that possibly a thing?

**Comment by atkinsj** [comment]
So after some research it appears that by default Ratchet falls back to a loop (i.e., while(true)) if libevent is unavailable. Libevent does not yet exist for PHP7.0, thus Ratchet is always going to use 100% CPU until that changes.

There is a fork of php-libevent at https://github.com/expressif/pecl-event-libevent but it generates a coredump for me on the pathfinder_websocket cmd.php.

I'm not sure of a good solution for this going forward.

**Comment by WildStrawberryEVE** [comment]
root@ap0:/var/log/pathfinder# sudo -u www-data  /usr/bin/php /var/www/pathfinder_websocket/cmd.php
2017-02-11 11:54:12 Server START ------------------------------------------
Segmentation fault (core dumped)
Feb 11 11:54:12 ap0 kernel: [1888276.414589] traps: php[24316] general protection ip:7ff4c9cde5d5 sp:7fff40eb01d0 error:0 in libevent.so[7ff4c9cda000+7000]

Got the same results :<
Osmanov's implementation fails a lot of tests on my system.

**Comment by atkinsj** [comment]
Yup. I'm not sure if this is a problem on Ratchet, react/event, pathfinder_websocket or the guys libevent fork either. I'll poke around with strace and see what I can hunt down tomorrow.

**Comment by exodus4d** [comment]
@atkinsj "*update client map  *" and "*update client user data*" are client side timings. Its just the time your web browser takes to update/sync your map  with updated data. 

- **UPDATE_CLIENT_MAP** -> time (ms) **after** your browser gets new map  data by Ajax **or** WebSocket **until** your browser finished updated the *DOM*
- **UPDATE_CLIENT_USER_DATA** -> time (ms) **after** your browser gets new **userData** (current location/system for all active characters on your current map) **until** your browser finished updating the *DOM*

The other 2 charts are **AJAX** timings. They show you how long your server takes to response:
- **UPDATE_SERVER_MAP** -> time (ms) between sending your updated map data (Ajax) (e.g. when you move a system around) to the server, **until** your server finished all the *DB* stuff and send the data back to you (and other characters, in case of WebSockets usw). -> Your browser takes the data up updates your map (see **UPDATE_CLIENT_MAP** )
- **UPDATE_SERVER_USER_DATA** -> time (ms) between sending a "ping" (Ajax) to your server. The server will receive that ping, and requests *CCPs* API for location updates for **your** character. (This is **not** effected by WebSocket configuration and is send every **5s**). The response of that Ajax call includes userData (current location) for all active characters on your map. -> Your browser takes the data up updates your map (see **UPDATE_CLIENT_USER_DATA** )

**Comment by atkinsj** [comment]
@exodus4d Thanks for the explanation on the variables.

The latest version with the recommended configuration is pretty much unusable. I'm running this on a 4 core 2.4GHz box with 4GB of ram. The Websocket server is using 100% CPU from the bug we mentioned above (lack of libevent support in php7.0) and redis-server is using 33-35%. The php-fpm process is also using a lot of CPU which makes me think that route calculations are not being cached but I haven't had a chance to confirm this.

This is using all latest development builds of things on a Ubuntu 16.04 box. Setting CACHE to folder= and disabling the websocket server returns back to a much more response Pathfinder overall and stops smashing the CPU.

**Comment by DanielMack** [comment]
just throwing a useless opinion around : never use any library or application which has `while(true)` in it - thats (very) bad programming, a beginners' mistake and a high-frequency problem source. Use atomics, volatile or something similar in your loop header and you're good to go - at least your app will not deadlock anymore. Just sayian.

**Comment by dkrotil** [comment]
@DanielMack we all know that, it would be fine if you can fix it ;)

**Comment by atkinsj** [comment]
The libevent library only has a while(true) as a fallback on systems where event polling isn't available.

Unfortunately PHP7 has no event polling framework available to it, putting us in this situation. That's one of the issues with using the bleeding edge.

**Comment by DanielMack** [comment]
sorry but ... that doest make much sense - you are using a programming mistake as **"fallback"** because you cannot use polling? Have you ever heard of something called "[busy waiting](https://en.wikipedia.org/wiki/Busy_waiting)"? Never use any deadlocking statement - under no circumstance, there is literally not a single use-case for it, even on embedded systems with extremely limited resources and policies

**Comment by atkinsj** [comment]
@DanielMack It does make sense and this isn't an issue with Pathfinder. Pathfinder uses Ratchet, a PHP library to provide WebSockets. That library uses pecl-lib-event which implements a variety of different polling and loop mechanisms (including inotify, libevent and a fallback to a while(true) when no system libraries are available). Your busy waiting Wikipedia link is completely irrelevant FWIW.

If you have issues with the implementation, take up with the Ratchet PHP project. Relevant issues are https://github.com/ratchetphp/Ratchet/issues/357 and https://github.com/reactphp/event-loop/issues/40.

**Comment by DanielMack** [comment]
thats like saying it makes sense to attack a wild bear with a dull stick because your favourite supermarket currently offers no lemonade.

**Comment by atkinsj** [comment]
You're welcome to submit a pull request to ratchetphp and reactphp with your no-system-library-required perfect event loop. Maybe leave this issue alone for now, though.

**Comment by DanielMack** [comment]
I'll probably do that sometime ... and it will get rejected because hobbyists dont like software dev reality, they make up their own.

**Comment by dkrotil** [comment]
@DanielMack there is no problem to create special fork like Daniel´s Ratchet for PF community.

**Comment by atkinsj** [comment]
So I think this is fixed upstream -- if you pull down the latest pathfinder_websocket project and then do a 'composer update', you'll pull down the new versions of React. The 100% CPU usage went away for me when this was done. @exodus4d

**Comment by atkinsj** [comment]
Ignore the comment I just deleted; I thought this had been fixed in upstream but the second a client connects it spikes to 100% CPU. It's unusable right now -- I have hundreds of users and even though I don't mind sacrificing a core to get WebSockets, it's not able to keep up with the users.

This would be a really cool feature if we could figure it out.

**Comment by DanielMack** [comment]
Well it'd be more of a bugfix, really ... i can confirm this, my instance is permanently spamming one core with a running websocket script.

**Comment by Tupsi** [comment]
I ended up removing websocket from our pf instance as well last week, because user where complaining about bad behavior. What is strange about it though is the fact, that they only started after I moved to the master 1.2 release. It was running just fine with the 100% one core hogging with an earlier 1.2.x dev version, but it might as well be that we just reached some critical userlevel to make it so unusable.
